website/docs/docs/building-a-dbt-project/building-models/python-models.md

```diff
@@ -0,0 +1,379 @@
+- [Snowpark Python: Creating UDFs](https://docs.snowflake.com/en/developer-guide/snowpark/python/creating-udfs.html)
+- ["PySpark functions: udf"](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.udf.html)
+
+:::info ❔ Our questions
+The process of creating and registering a "named" UDF is different across data platforms. Snowpark also supports ["vectorized" UDFs](https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-batch.html): Pandas-like functions that can be executed in parallel.
+
+- Should dbt have a role in abstracting over UDFs? Should dbt support a new type of DAG node, `function`?
+- Is the primary motivation for code reuse across Python models, or defining Python-language functions that can be called from SQL models?
+:::
+
+### Data frame API + syntax
+
+Over the past decade, most people writing data transformations in Python have adopted <Term id="dataframe">data frame</Term> as their common abstraction. dbt follows this convention by returning `ref()` and `source()` as data frames, and it expects all Python models to return a data frame.
+
+A data frame is a two dimensional data structure (rows and columns). It supports convenient methods for transforming that data, creating new columns from calculations performed on existing columns. It also offers convenient methods for previewing data while developing locally or in a notebook.
+
+That's about where the agreement ends. There are numerous frameworks with their own syntaxes and APIs for data frames. The [Pandas](https://pandas.pydata.org/docs/) library offered one of the original data frame APIs, and its syntax is the most common to learn for new data professionals. Most newer data frame APIs offer some degree of compatibility with Pandas-style syntax, though few can offer perfect interoperability. This is true for both Snowpark and PySpark, which have their own DataFrame APIs.
+
+When developing a Python model, you will find yourself asking these questions:
+
+**Why Pandas?** The most common API for data frames. Makes it easy to explore sampled data and develop transformations locally. For small datasets, you can "promote" your code as-is into dbt models, and run in production.
+
+**Why _not_ Pandas?** Performance. Pandas runs "single-node" transformations, which cannot benefit from the parallelism and distributed compute offered by modern data warehouses. This quickly becomes a problem as you operate on larger datasets. Some data platforms do support optimizations for code written using Pandas' data frame API, preventing the need for major refactors. For example, ["Pandas on PySpark"](https://spark.apache.org/docs/latest/api/python/getting_started/quickstart_ps.html) offers support for 95% of Pandas functionality using the same API.
```

Comment:
"offers support for 95% of Pandas functionality using the same API" => "offers support for 95% of Pandas functionality using the same API while still leveraging parallel processing"